File: AnkiDroid/src/test/java/com/wildplot/android/rendering/PieChartTestParametersBuilder.java
//noinspection MissingCopyrightHeader #8659

package com.wildplot.android.rendering;


import com.wildplot.android.rendering.graphics.wrapper.ColorWrap;

import static java.lang.String.format;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

final class PieChartTestParametersBuilder {
    private final double[] mValues;
    private final int mNumberOfValues;
    private final double mSum;
    private final double mFirstSectorAngle;
    private final double[] mStartAngles;
    private final double[] mArcLengths;
    private final ColorWrap[] mColors;

    PieChartTestParametersBuilder(double[] values,
                                  double firstSectorAngle) {
        if (values.length < 1) {
            throw new IllegalArgumentException("Empty array of values");
        }
        mValues = values;
        mNumberOfValues = values.length;
        mSum = calcSum(values);
        if (mSum == 0) {
            throw new IllegalArgumentException(
                    format("All %d values are zero", values.length));
        }
        mFirstSectorAngle = firstSectorAngle;
        mStartAngles = new double[mNumberOfValues];
        mArcLengths = new double[mNumberOfValues];
        mColors = new ColorWrap[mNumberOfValues];
        calcArcLengths();
        calcStartAngles();
        fillColors();
    }

    private double calcSum(double[] values) {
        double sum = 0.;
        for (double v : values) {
            sum += v;
        }
        return sum;
    }

    private void calcArcLengths() {
        for (int i = 0; i < mNumberOfValues; i++) {
            mArcLengths[i] = 360.0 * mValues[i] / mSum;
        }
    }

    private void calcStartAngles() {
        mStartAngles[0] = mFirstSectorAngle;
        for (int i = 1; i < mNumberOfValues; i++) {
            mStartAngles[i] = mStartAngles[i - 1] + mArcLengths[i - 1];
        }
    }

    private void fillColors() {
        for (int i = 0; i < mNumberOfValues; i++) {
            mColors[i] = createColorMock(i);
        }
    }

    private ColorWrap createColorMock(int i) {
        ColorWrap c = mock(ColorWrap.class);
        when(c.getColorValue()).thenReturn(i);
        return c;
    }

    double[] getStartAngles() {
        return mStartAngles;
    }

    double[] getArcLengths() {
        return mArcLengths;
    }

    ColorWrap[] getColors() {
        return mColors;
    }
}

File: AnkiDroid/src/test/java/com/wildplot/android/rendering/PieChartTestParametersBuilder.kt
//noinspection MissingCopyrightHeader #8659
package com.wildplot.android.rendering

import com.wildplot.android.rendering.graphics.wrapper.ColorWrap
import org.mockito.Mockito.mock
import org.mockito.kotlin.whenever

internal class PieChartTestParametersBuilder(
    values: DoubleArray,
    firstSectorAngle: Double
) {
    private val mValues: DoubleArray = values
    private val mNumberOfValues: Int = values.size
    private val mSum: Double = calcSum(values)
    private val mFirstSectorAngle: Double = firstSectorAngle
    val startAngles: DoubleArray = DoubleArray(mNumberOfValues)
    val arcLengths: DoubleArray = DoubleArray(mNumberOfValues)
    val colors: Array<ColorWrap?> = arrayOfNulls(mNumberOfValues)
    private fun calcSum(values: DoubleArray): Double {
        var sum = 0.0
        for (v in values) {
            sum += v
        }
        return sum
    }

    private fun calcArcLengths() {
        for (i in 0 until mNumberOfValues) {
            arcLengths[i] = 360.0 * mValues[i] / mSum
        }
    }

    private fun calcStartAngles() {
        startAngles[0] = mFirstSectorAngle
        for (i in 1 until mNumberOfValues) {
            startAngles[i] = startAngles[i - 1] + arcLengths[i - 1]
        }
    }

    private fun fillColors() {
        for (i in 0 until mNumberOfValues) {
            colors[i] = createColorMock(i)
        }
    }

    private fun createColorMock(i: Int): ColorWrap {
        val c = mock(ColorWrap::class.java)
        whenever(c.colorValue).thenReturn(i)
        return c
    }

    init {
        require(values.isNotEmpty()) { "Empty array of values" }
        require(mSum != 0.0) { "All ${values.size} values are zero" }
        calcArcLengths()
        calcStartAngles()
        fillColors()
    }
}
